src/NHibernate/Action/DelayedPostInsertIdentifier.cs

```diff
@@ -15,20 +16,27 @@ namespace NHibernate.Action
 	[Serializable]
 	public class DelayedPostInsertIdentifier
 	{
-		[ThreadStatic]
-		private static long _Sequence = 0;
+		private static long GlobalSequence = 0;
 		private readonly long sequence;
 
 		public DelayedPostInsertIdentifier()
 		{
-			lock (typeof(DelayedPostInsertIdentifier))
+			sequence = GetNextSequence();
+		}
+
+		private static long GetNextSequence()
+		{
+			long initialValue, computedValue;
+			do
 			{
-				if (_Sequence == long.MaxValue)
-				{
-					_Sequence = 0;
-				}
-				sequence = _Sequence++;
-			}
+				initialValue = GlobalSequence;
+				computedValue =
+					initialValue == long.MaxValue
+						? 1
+						: initialValue + 1;
+			} while (Interlocked.CompareExchange(ref GlobalSequence, computedValue, initialValue) != initialValue);
+
+			return computedValue;
```

Comment:
Why not just `Interlocked.Increase(ref GlobalSequence)`?

Reply:
To avoid negative and zero values in keys.

Reply:
I don't want how to implement counter reset with `Interlocked.Increase`'. I mean  to make counter after `long.MaxValue` generate 1.
Negative and zero values might not be the issue  -  but as it's public class I decided to keep old behavior.

Reply:
On the other hand it's only exposed in `ToString`. So let me know if you think it's OK to have negative/zero sequence field value and I will change it to `Interlocked.Increment`

Reply:
I don't mind negative "sequences". If you don't like negative numbers in this context we can make the `sequence` of type `uint` or `ulong`.

In fact, I don't think we need _any_ sequence there: we could remove all fields from this class and use reference comparison only.

Reply:
There is no `Interlocked.Increment` for unsigned types. And as this type is serialize I don't want to run into some strange issues due to inequality of deserialized keys because of reference comparison.
So let's simply use `Interlocked.Increment`

Reply:
Ok.